{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "146bb1c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Loading in packages-----#\n",
    "import scvi\n",
    "import anndata as ad #building anndata objects\n",
    "import scanpy as sc # Single cell analysis \n",
    "import pandas as pd # dataframe manipulation \n",
    "import numpy as np\n",
    "from scipy.io import mmread #Reading in mtx file \n",
    "from scipy.sparse import csr_matrix #convert from coo to csr matrix\n",
    "from rich import print\n",
    "import seaborn as sns\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8ee39e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Setting Seed-----#\n",
    "scvi.settings.seed = 0\n",
    "print(\"Last run with scvi-tools version:\", scvi.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc4fa816",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Generating Anndata from Seurat Object-----#\n",
    "conversion_dir = \"/projects/b1217/HHA/Multiome_Scanpy_Conversion/Full_Atlas/\"\n",
    "#Contains metadata, barcodes, and umap embeddings \n",
    "metadata = pd.read_csv(os.path.join(conversion_dir, \"SCRNA_Multiome_IFE_Metadata_4_1_25.csv\"))\n",
    "#Reading in the count matrix from seurat.combined\n",
    "counts = mmread(os.path.join(conversion_dir, \"SCRNA_Multiome_IFE_RawCounts_4_1_25.mtx\"))\n",
    "counts = csr_matrix(counts.transpose()) #transposing to cell x gene matrix and converting to csr format\n",
    "#Contains gene names \n",
    "gene_names = pd.read_csv(os.path.join(conversion_dir, \"SCRNA_Multiome_IFE_GeneNames_4_1_25.csv\"))\n",
    "#Log normalized expression for stress genes\n",
    "stress_genes = pd.read_csv(os.path.join(conversion_dir, \"SCRNA_Multiome_IFE_StressSig_LogNorm_4_1_25.csv\"))\n",
    "stress_genes = stress_genes.set_index(\"Barcode\")\n",
    "#Log normalized expression for cell cycle genes\n",
    "cc_genes = pd.read_csv(os.path.join(conversion_dir, \"SCRNA_Multiome_IFE_CellCycle_LogNorm_4_1_25.csv\"))\n",
    "cc_genes = cc_genes.set_index(\"Barcode\")\n",
    "#Converting to anndata object\n",
    "HHA = ad.AnnData(counts) #creating anndata object\n",
    "HHA.obs_names = metadata[\"barcode\"] #Adding barcodes as obs names \n",
    "HHA.var_names = gene_names[\"gene\"] #adding gene names \n",
    "#Saving Count Layer\n",
    "HHA.layers[\"counts\"] = HHA.X.copy()\n",
    "#Viewing results\n",
    "print(HHA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d9cdb1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Adding metadata to anndata--------#\n",
    "HHA.obs = metadata\n",
    "HHA.obs_names = metadata[\"barcode\"]\n",
    "#Adding normalized cell cycle gene expression to obs\n",
    "HHA.obs = pd.concat([HHA.obs, cc_genes], axis =1)\n",
    "#Adding normalized stress gene expression to obs\n",
    "HHA.obs = pd.concat([HHA.obs, stress_genes], axis =1)\n",
    "HHA.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0925962",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Setting Up SCVI Model-----#\n",
    "#Specifying the donor and sequencing platform as additional categorical covariates for integration\n",
    "#Not including mitochondrial and ribosomal percentage or cell cycle as covariates, as this may mask real biology\n",
    "continuous_covariates = stress_genes.columns.to_list() + cc_genes.columns.to_list() + [\"nCount_RNA\", \"nFeature_RNA\", \"percent.mt\", \"percent.ribo\"]\n",
    "scvi.model.SCVI.setup_anndata(HHA, layer=\"counts\", \n",
    "                              batch_key=\"SampleID\",\n",
    "                              categorical_covariate_keys = [\"DonorID\", \"Sex\", \"Platform\"],\n",
    "                              continuous_covariate_keys = continuous_covariates)\n",
    "model = scvi.model.SCVI(HHA, n_layers=2, n_latent=50, gene_likelihood= 'nb')\n",
    "# view anndata_setup as a sanity check\n",
    "model.view_anndata_setup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47366728",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Training model using gpu acceleration\n",
    "model.train(max_epochs = 500, early_stopping = True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "602ca592",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Plotting Model convergence-----#\n",
    "train_elbo = model.history[\"elbo_train\"][1:]\n",
    "test_elbo = model.history[\"elbo_validation\"]\n",
    "ax = train_elbo.plot()\n",
    "test_elbo.plot(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d77be5f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Extracting Latent Representation from Model----#\n",
    "HHA.obsm[\"X_scVI\"] = model.get_latent_representation()\n",
    "HHA.obsm[\"X_scVI\"]\n",
    "\n",
    "#----Leiden Clustering with default Scanpy workflow------#\n",
    "sc.pp.neighbors(HHA, use_rep= \"X_scVI\")\n",
    "sc.tl.leiden(HHA)\n",
    "sc.tl.umap(HHA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9ff80cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#----- Quickly Visualizing results with UMAP embeddding------#\n",
    "sc.pl.embedding(\n",
    "    HHA,\n",
    "    basis=\"X_umap\",\n",
    "    color=['GlobalAnnotation', \"SampleID\", 'Sex', \"Phase\", 'leiden'],\n",
    "    frameon=False,\n",
    "    add_outline=True,\n",
    "    ncols=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd2279a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Checking QC Metric mixing in UMAP embeddings------#\n",
    "sc.pl.embedding(HHA, basis=\"X_umap\",\n",
    "    color=[\"nCount_RNA\", \"nFeature_RNA\", \"percent.mt\", \"percent.ribo\"],\n",
    "    cmap = \"plasma\", frameon=False, add_outline=True, size = 7.5, ncols=2)\n",
    "sc.pl.embedding(HHA, basis=\"X_umap\",\n",
    "    color=[\"DissociationScore\"],\n",
    "    cmap = \"turbo\", add_outline=True, size = 7.5, frameon=False, ncols=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5301d08b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------Extracting SCVI Latent Dimensions and Writing to CSV-----#\n",
    "HHA_Latent = pd.DataFrame(HHA.obsm[\"X_scVI\"])\n",
    "HHA_Latent.columns = [\"SCVI_\" + str(i) for i in range(1, 51)]\n",
    "HHA_Latent[\"Barcode\"] = HHA.obs_names\n",
    "HHA_Latent.head(10)\n",
    "HHA_Latent.to_csv(os.path.join(conversion_dir, \"HHA_SCRNA_Multiome_RNA_IFE_LatentRep_4_1_25.csv\"),\n",
    "                               index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b08e3c9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#----Rerunning model after filtering Fos/Jun outliers-----#\n",
    "filtered_barcodes = pd.read_csv(os.path.join(conversion_dir, \"HHA_IFE_Stress_Filtered_Barcodes_4_1_25.csv\"))\n",
    "HHA_Filtered = HHA[HHA.obs[\"barcode\"].isin(filtered_barcodes[\"Barcode\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc10b1be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Setting Up SCVI Model-----#\n",
    "#Specifying the donor and sequencing platform as additional categorical covariates for integration\n",
    "#Not including mitochondrial and ribosomal percentage or cell cycle as covariates, as this may mask real biology\n",
    "continuous_covariates = stress_genes.columns.to_list() + cc_genes.columns.to_list() + [\"nCount_RNA\", \"nFeature_RNA\", \"percent.mt\", \"percent.ribo\"]\n",
    "scvi.model.SCVI.setup_anndata(HHA_Filtered, layer=\"counts\", \n",
    "                              batch_key=\"SampleID\",\n",
    "                              categorical_covariate_keys = [\"DonorID\", \"Sex\", \"Platform\"],\n",
    "                              continuous_covariate_keys = continuous_covariates)\n",
    "filtered_model = scvi.model.SCVI(HHA_Filtered, n_layers=2, n_latent=50, gene_likelihood= 'nb')\n",
    "# view anndata_setup as a sanity check\n",
    "filtered_model.view_anndata_setup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "797c77af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Training model using gpu acceleration\n",
    "filtered_model.train(max_epochs = 500, early_stopping = True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89fd5e08",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Plotting Model convergence-----#\n",
    "train_elbo_filtered = filtered_model.history[\"elbo_train\"][1:]\n",
    "test_elbo_filtered = filtered_model.history[\"elbo_validation\"]\n",
    "ax = train_elbo_filtered.plot()\n",
    "test_elbo_filtered.plot(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80079bf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Extracting Latent Representation from Model----#\n",
    "HHA_Filtered.obsm[\"X_scVI\"] = filtered_model.get_latent_representation()\n",
    "HHA_Filtered.obsm[\"X_scVI\"]\n",
    "\n",
    "#----Leiden Clustering with default Scanpy workflow------#\n",
    "sc.pp.neighbors(HHA_Filtered, use_rep= \"X_scVI\")\n",
    "sc.tl.leiden(HHA_Filtered)\n",
    "sc.tl.umap(HHA_Filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63a8d159",
   "metadata": {},
   "outputs": [],
   "source": [
    "#----- Quickly Visualizing results with UMAP embeddding------#\n",
    "sc.pl.embedding(\n",
    "    HHA_Filtered,\n",
    "    basis=\"X_umap\",\n",
    "    color=['GlobalAnnotation', \"SampleID\", 'Sex', \"Phase\", 'leiden'],\n",
    "    frameon=False,\n",
    "    add_outline=True,\n",
    "    ncols=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccfbe615",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Checking QC Metric mixing in UMAP embeddings------#\n",
    "sc.pl.embedding(HHA_Filtered, basis=\"X_umap\",\n",
    "    color=[\"nCount_RNA\", \"nFeature_RNA\", \"percent.mt\", \"percent.ribo\"],\n",
    "    cmap = \"plasma\", frameon=False, add_outline=True, size = 7.5, ncols=2)\n",
    "sc.pl.embedding(HHA_Filtered, basis=\"X_umap\",\n",
    "    color=[\"DissociationScore\"],\n",
    "    cmap = \"turbo\", add_outline=True, size = 7.5, frameon=False, ncols=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ee00c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Checking on gene expression-----#\n",
    "#Normalizing genes\n",
    "sc.pp.normalize_total(HHA_Filtered)\n",
    "sc.pp.log1p(HHA_Filtered)\n",
    "#Plotting on UMAP\n",
    "sc.pl.embedding(HHA_Filtered, basis=\"X_umap\",\n",
    "    color=[\"COL17A1\", \"KRT5\", \"KRT10\", \"KRT1\", \"KRTDAP\", \"IVL\", \"GATA3\", \"SOX6\", \"FOS\"],\n",
    "                add_outline=True, size = 15,\n",
    "    cmap = \"plasma\", frameon=False, ncols=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d8acec9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------Extracting SCVI Latent Dimensions and Writing to CSV-----#\n",
    "HHA_Latent_Filtered = pd.DataFrame(HHA_Filtered.obsm[\"X_scVI\"])\n",
    "HHA_Latent_Filtered.columns = [\"SCVI_\" + str(i) for i in range(1, 51)]\n",
    "HHA_Latent_Filtered[\"Barcode\"] = HHA_Filtered.obs_names\n",
    "HHA_Latent_Filtered.head(10)\n",
    "HHA_Latent_Filtered.to_csv(os.path.join(conversion_dir, \"HHA_SCRNA_Multiome_RNA_IFE_PostFiltering_LatentRep_4_1_25.csv\"),\n",
    "                               index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd307acb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SCVI SpatialData Kernel (SCVIEnv)",
   "language": "python",
   "name": "scvienv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
