{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81606fad",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------Loading in Packages------#\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "from scipy.io import mmread\n",
    "from scipy.sparse import csr_matrix\n",
    "import seaborn as sns\n",
    "import os\n",
    "import tacco as tc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1d77e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Loading in 8um bin filtered dataset-------#\n",
    "adata = sc.read_h5ad(\"/projects/b1217/HHA/EL_S1_adata/EL_S1_full_sdata8um-f100.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a37acb46",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------Creating EnsemblID column and set as index--------#\n",
    "adata.var['ensembl'] = adata.var['gene_ids'].copy()\n",
    "adata = adata[:, adata.var[\"ensembl\"] != \"Unknown\"].copy()\n",
    "new_index = adata.var[\"ensembl\"]\n",
    "adata.var.index = new_index\n",
    "adata.var_names = new_index\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0f57683",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------Loading in Full scRNA Atlas-----#\n",
    "HHA = sc.read_h5ad(\"/projects/b1217/Edward/R_Projects/HHA/h5ad/20250516_HHA-SC-Final.h5ad\")\n",
    "HHA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9c0d42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------Plotting UMAPs------#\n",
    "#General Annotation\n",
    "sc.pl.umap(HHA, color=\"GeneralAnnotation\", frameon=False, legend_loc=\"on data\")\n",
    "#Final Annotation\n",
    "sc.pl.umap(HHA, color=\"FinalAnnotation\", frameon=False, legend_loc=\"on data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "965edad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Printing Cell Counts for each Annotation------#\n",
    "for cat, n in HHA.obs[\"FinalAnnotation\"].value_counts().items():\n",
    "    print(f\"{cat}: {n}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53485e85",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------Loading in Matrix scRNA-seq AnnData-------#\n",
    "Matrix = sc.read_h5ad(\"/projects/b1217/HHA/Bulb_Recluster_5_22_AnnData/Matrix_Palantir_6_11_25.h5ad\")\n",
    "#create new column in var as names\n",
    "Matrix.var[\"symbol\"] = Matrix.var.index.copy()\n",
    "Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76411d53",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------Color Dictionaries------#\n",
    "#Colors for Fine Annotation\n",
    "fine_colors = {\"Lower COL17\": \"#89C75F\",\n",
    "               \"Upper COL17\": \"#3BBCA8\",\n",
    "               \"LPC\": \"#208A42\",\n",
    "               \"Medulla\": \"#0C727C\",\n",
    "               \"Early_Cortex\": \"#9ECAE1\",\n",
    "               \"Middle_Cortex\": \"#4292C6\",\n",
    "               \"Late_Cortex\": \"#08306B\",\n",
    "               \"Early_Cuticle\": \"#E6C2DC\",\n",
    "               \"Middle_Cuticle\": \"#C06CAB\",\n",
    "               \"Late_Cuticle\": \"#89288F\",\n",
    "               \"Early_IRS_I\": \"#D8A767\",\n",
    "               \"Early_IRS_II\": \"#F47D2B\",\n",
    "               \"IRS_Henle\": \"#F37B7D\",\n",
    "               \"IRS_Huxley\": \"#7E1416\",\n",
    "               \"IRS_Cuticle\": \"#D24B27\"}\n",
    "#Colors for Broad Annotation\n",
    "broad_colors = {\"COL17\": \"#89C75F\",\n",
    "               \"LPC\": \"#208A42\",\n",
    "               \"Cortex\": \"#4292C6\",\n",
    "               \"Cuticle\": \"#89288F\",\n",
    "               \"IRS\": \"#D24B27\"}\n",
    "\n",
    "#--------Plotting UMAPs------#\n",
    "#Fine Annotation\n",
    "sc.pl.umap(Matrix, color=\"MatrixAnnotationFine\", palette = fine_colors,\n",
    "    frameon=False, size = 15, legend_loc=\"on data\")\n",
    "#Broad Annotation\n",
    "sc.pl.umap(Matrix, color=\"MatrixAnnotationBroad\", palette = broad_colors,\n",
    "           legend_loc=\"on data\", frameon = False, size = 15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a29d47aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Crossmapping var_names to EnsemblID--------#\n",
    "symbol_to_ensembl = dict(zip(HHA.var[\"names\"], HHA.var[\"ensembl\"]))\n",
    "def map_to_ensembl(name):\n",
    "    if name.startswith(\"ENSG\"):\n",
    "        return name\n",
    "    return symbol_to_ensembl.get(name, \"Unknown\")\n",
    "Matrix.var[\"ensembl\"] = Matrix.var[\"symbol\"].apply(map_to_ensembl)\n",
    "\n",
    "#-------Creating EnsemblID column and set as index--------#\n",
    "Matrix = Matrix[:, Matrix.var[\"ensembl\"] != \"Unknown\"].copy()\n",
    "new_index = Matrix.var[\"ensembl\"]\n",
    "Matrix.var.index = new_index\n",
    "Matrix.var_names = new_index\n",
    "Matrix.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54a60269",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Loading in Mask for QC-------#\n",
    "#Contains whether cells were filtered during QC selection during Matrix subclustering or are in the matrix. \n",
    "QC_Metadata = pd.read_csv(\"/projects/b1217/HHA/Multiome_Seurat/HHA_Full_Matrix_Mask_5_30_25.csv\")\n",
    "QC_Metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16456119",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Filtering for Cells retained after Additional Matrix QC-------#\n",
    "QC_Mask = QC_Metadata[\"Barcode\"][QC_Metadata[\"QCFiltered\"] == True].astype(\"str\")\n",
    "HHA = HHA[~HHA.obs_names.isin(QC_Mask)].copy()\n",
    "HHA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71768606",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------Mapping Matrix Annotations to Filtered HHA--------#\n",
    "#All non-matrix cells labeled Unknown\n",
    "HHA.obs.index = HHA.obs.index.astype(str)\n",
    "Matrix.obs.index = Matrix.obs.index.astype(str)\n",
    "Matrix_Mapping = Matrix.obs[\"MatrixAnnotationFine\"].to_dict()\n",
    "HHA.obs[\"MatrixAnnotation\"] = HHA.obs.index.to_series().map(Matrix_Mapping).fillna(\"Other\")\n",
    "#Adding new Annotation with Matrix + Melanocytes\n",
    "#Pulling melanocyte indices and creating dictionary\n",
    "Mel = HHA[HHA.obs[\"FinalAnnotation\"].isin([\"Melanocytes I\", \"Melanocytes II\"]),:].copy()\n",
    "Mel_Mapping = Mel.obs[\"FinalAnnotation\"].to_dict()\n",
    "#Merging dictionaries \n",
    "MelMat_Mapping = Matrix_Mapping | Mel_Mapping\n",
    "#Adding to Object\n",
    "HHA.obs[\"MelMatAnnotation\"] = HHA.obs.index.to_series().map(MelMat_Mapping).fillna(\"Other\")\n",
    "\n",
    "#--------Checking Labels-------#\n",
    "sc.pl.umap(HHA, color=\"MelMatAnnotation\", frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "661c5ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Printing Cell Counts for each Annotation------#\n",
    "for cat, n in HHA.obs[\"MelMatAnnotation\"].value_counts().items():\n",
    "    print(f\"{cat}: {n}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f35fae6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------Converting gene expression matrices to float for TACCO------#\n",
    "HHA.X = HHA.X.astype(np.float64)\n",
    "adata.X = adata.X.astype(np.float64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a659bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Running TACCO on Full Dataset-------#\n",
    "tc.tl.annotate(adata, HHA, annotation_key='MelMatAnnotation', result_key='MelMatAnnotation')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d54d5188",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Retrieving Best Annotation Mapping for Each Cluster------#\n",
    "tc.utils.get_maximum_annotation(adata, 'MelMatAnnotation', result_key='MelMatAnnotation')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "778a592a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------Plotting Initial TACCO Results-----#\n",
    "#mapping unknown to transparent\n",
    "matrix_colors_spatial = {\"Lower COL17\": \"#89C75F\",\n",
    "               \"Upper COL17\": \"#3BBCA8\",\n",
    "               \"LPC\": \"#208A42\",\n",
    "               \"Medulla\": \"#0C727C\",\n",
    "               \"Early_Cortex\": \"#9ECAE1\",\n",
    "               \"Middle_Cortex\": \"#4292C6\",\n",
    "               \"Late_Cortex\": \"#08306B\",\n",
    "               \"Early_Cuticle\": \"#E6C2DC\",\n",
    "               \"Middle_Cuticle\": \"#C06CAB\",\n",
    "               \"Late_Cuticle\": \"#89288F\",\n",
    "               \"Early_IRS_I\": \"#D8A767\",\n",
    "               \"Early_IRS_II\": \"#F47D2B\",\n",
    "               \"IRS_Henle\": \"#F37B7D\",\n",
    "               \"IRS_Huxley\": \"#7E1416\",\n",
    "               \"IRS_Cuticle\": \"#D24B27\",\n",
    "               \"Melanocytes I\": \"#916848\", \n",
    "               \"Melanocytes II\": \"#722A2D\",\n",
    "               \"Other\": \"#00000000\"}\n",
    "\n",
    "#--------Plotting Full Image-------#\n",
    "fig, ax = plt.subplots(figsize=(40, 40))\n",
    "sc.pl.spatial(adata, color=[\"MelMatAnnotation\"], img_key=\"hires\", ax=ax, palette = matrix_colors_spatial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "102fb5a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------Second Iteration TACCO Run on spots mapped to Matrix/Melanocytes-------#\n",
    "#Filtering Down to Matrix/Melanocyte Hits\n",
    "adata_sub = adata[~adata.obs[\"MelMatAnnotation\"].isin([\"Other\"])].copy()\n",
    "print(adata_sub)\n",
    "#Filtering scRNA for Matrix, Melanocytes \n",
    "MelMat = HHA[~HHA.obs[\"MelMatAnnotation\"].isin([\"Other\"])].copy()\n",
    "#Remapping Categories to exclude other for TACCO\n",
    "MelMat.obs[\"MelMatAnnotationFine\"] = pd.Categorical(MelMat.obs[\"MelMatAnnotation\"],\n",
    "                                                   categories = ['Early_Cortex', 'Early_Cuticle', 'Early_IRS_I', 'Early_IRS_II',\n",
    "       'IRS_Cuticle', 'IRS_Henle', 'IRS_Huxley', 'LPC', 'Late_Cortex',\n",
    "       'Late_Cuticle', 'Lower COL17', 'Medulla', 'Melanocytes I',\n",
    "       'Melanocytes II', 'Middle_Cortex', 'Middle_Cuticle', 'Upper COL17'])\n",
    "print(MelMat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0b85e77",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Rerunning Annotation: Matrix to all Matrix/Melanocyte Bins------#\n",
    "tc.tl.annotate(adata_sub, Matrix, annotation_key='MatrixAnnotationFine', result_key='MatrixAnnotationFine')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec4b3d13",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Retrieving Best Annotation Mapping for Each Cluster------#\n",
    "tc.utils.get_maximum_annotation(adata_sub, 'MatrixAnnotationFine', result_key='MatrixAnnotationFine')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f280f6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------Plotting Matrix Mappings-------#\n",
    "fig, ax = plt.subplots(figsize=(40, 40))\n",
    "sc.pl.spatial(adata_sub, color=[\"MatrixAnnotationFine\"], img_key=\"hires\", ax=ax, palette = matrix_colors_spatial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "440fa907",
   "metadata": {},
   "outputs": [],
   "source": [
    "#---------Mapping COL17 cells along DP border------------#\n",
    "adata_col17 = adata[adata.obs[\"MelMatAnnotation\"].isin(\n",
    "    [\"Melanocytes I\", \"Melanocytes II\", \"Upper COL17\", \"Lower COL17\", \"Medulla\", \"LPC\"])].copy()\n",
    "\n",
    "#--------Plotting Full Image-------#\n",
    "fig, ax = plt.subplots(figsize=(40, 40))\n",
    "sc.pl.spatial(adata_col17, color=[\"MelMatAnnotation\"], img_key=\"hires\", ax=ax, palette = matrix_colors_spatial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4952a2fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------Subsetting for COL17A1 Populations-----------#\n",
    "COL17 = Matrix[Matrix.obs[\"MatrixAnnotationFine\"].isin([\"Upper COL17\", \"Lower COL17\", \"Medulla\", \"LPC\"])].copy()\n",
    "#Removing other categories for TACCO Mapping\n",
    "COL17.obs[\"COL17Annotation\"] = pd.Categorical(COL17.obs[\"MatrixAnnotationFine\"],\n",
    "                                                   categories = [\"Upper COL17\", \"Lower COL17\", \"Medulla\", \"LPC\"])\n",
    "COL17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7218a30",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Printing Cell Counts for each Annotation------#\n",
    "for cat, n in COL17.obs[\"COL17Annotation\"].value_counts().items():\n",
    "    print(f\"{cat}: {n}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7640f0d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#------Rerunning Annotation: Matrix to all Matrix/Melanocyte Bins------#\n",
    "tc.tl.annotate(adata_col17, COL17, annotation_key='COL17Annotation', result_key='COL17Annotation')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b64c3145",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----Retrieving Best Annotation Mapping for Each Cluster------#\n",
    "tc.utils.get_maximum_annotation(adata_col17, 'COL17Annotation', result_key='COL17Annotation')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "303eeaff",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------Plotting Full Image-------#\n",
    "fig, ax = plt.subplots(figsize=(40, 40))\n",
    "sc.pl.spatial(adata_col17, color=[\"COL17Annotation\"], img_key=\"hires\", ax=ax, palette = matrix_colors_spatial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f72b00aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#---------------Pulling Scores from Obsm----------------#\n",
    "Annot_Scores = adata_col17.obsm['COL17Annotation'].copy()\n",
    "Annot_Names = [x.replace(\" \", \"_\") + \"_Score\" for x in Annot_Scores.columns]\n",
    "Annot_Scores.columns = Annot_Names\n",
    "Annot_Scores\n",
    "#Adding to obs\n",
    "for Score in Annot_Scores.columns:\n",
    "    adata_col17.obs[Score] = Annot_Scores[Score]\n",
    "adata_col17.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a742613b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#----------------Saving Object------------#\n",
    "adata_col17.write_h5ad(\"/projects/b1217/HHA/Bulb_Spatial/HHA_Spatial_EL_S1_TACCO_COL17_Mapping_12_15_25.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0d154e7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TACCO Mapping Kernel (TACCOEnv)",
   "language": "python",
   "name": "taccoenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
